documentation/Calls.md

```diff
@@ -0,0 +1,636 @@
+and ranks is always a compilation-time task (15.5.6, C.10.6)).
+
+Each derived type that declares or inherits any overridable
+type-bound procedure bindings must correspond to a static constant
+table of code addresses (or, more likely, a static constant type
+description containing or pointing to such a table, along with
+information used by the runtime support library for initialization,
+copying, finalization, and I/O of type instances).  Each overridable
+type-bound procedure in the type corresponds to an index into this table.
+
+### Host instance linkage
+
+Calls to dummy procedures and procedure pointers that resolve to
+internal procedures need to pass an additional "host instance" argument that
+addresses a block of storage in the stack frame of the their
+host subprogram that was active at the time they were passed as an
+effective argument or associated with a procedure pointer.
+This is similar to a static link in implementations of programming
+languages with nested subprograms, although Fortran only allows
+one level of nesting.
+The 64-bit x86 and little-endian OpenPower ABIs reserve registers
+for this purpose (`%r10` & `R11`); 64-bit ARM has a reserved register
+that can be used (`x18`).
+
+The host subprogram objects that are visible to any of their internal
+subprograms need to be resident in memory across any calls to them
+(direct or not).  Any host subprogram object that might be defined
+during a call to an internal subprogram needs to be reloaded after
+a call or reside permanently in memory.
+A simple conservative analysis of the internal subprograms can
+identify all of these escaping objects and their definable subset.
+
+The address of the host subprogram storage used to hold the escaping
+objects needs to be saved alongside the code address(es) that
+represent a procedure pointer.
+It also needs to be conveyed alongside the text address for a
+dummy procedure.
+
+For F77ish procedures, we cannot use a "procedure pointer descriptor"
+to pass a procedure argument -- they expect to receive a single
+address argument.
+We will need to package the host instance link in a trampoline
+that loads its address into the designated register.
+
+GNU Fortran and Intel Fortran construct trampolines by writing
+a sequence of machine instructions to a block of storage in the
+host's stack frame, which requires the stack to be executable,
+which seems inadvisable for security reasons;
+XLF manages trampolines in its runtime support library, which adds some overhead
+to their construction and a reclamation obligation;
+NAG Fortran manages a static fixed-sized stack of trampolines
+per call site, imposing a hidden limit on recursion and foregoing
+reentrancy;
+PGI passes host instance links in descriptors in additional arguments
+that are not always successfully forwarded across implicit interfaces,
+sometimes leading to crashes when they turn out to be needed.
+
+F18 will manage a pool of trampolines in its runtime support library
```

Comment:
I am not sure I fully understand the concept of trampoline.
Would that be something like:
in the host function
```func_pointer => SetTrampoline(proc, host_link)```

In SetTrampoline:
```
N = GetFreeTrampoline()
host_links[N] = host_link
proc[N] = proc
return trampolines[N] // trampoline<N>
```

In the trampoline function `trampoline<N>`
```
host_register = host_links[N]
return proc[N]()
```

Reply:
OK, so it would not be a pre-compiled pool, more like a JIT compiled one.
I did not know it was safer than stack execution, but after reading on the topic a bit, I guess you want to use something like W^X protection (http://man7.org/linux/man-pages/man2/mprotect.2.html). Sounds great to allow deeper recursions.